content/posts/redis-modules-into-arm-land-part1.md

```diff
@@ -0,0 +1,274 @@
+In this series of posts, I'll describe our vision of ARM platform support and the developer user experience, as well as the steps we took along the path, as the way was as valuable as the outcome.
+
+If you'll follow through, you'll end up with a fully functional ARM build laboratory.
+
+## Inside RedisEdge
+
+![redisedge1](/redis-modules-arm-1.png)
+
+But first, let's take a look at RedisEdge.
+RedisEdge is not a Redis module, but an aggregate of three Redis modules: RedisGears, RedisAI, and RedisTimeSeries.
+It is distributed as a Docker image, which is based on Redis Server 5.0. Thus, one can simply pull the image, run it, and start issuing Docker commands, load models into RedisAI, and execute Python gears scripts on RedisGears.
+Although one can easily get Docker out of the equation (by installing a Redis server and copying Redis Modules files), we'll see that Docker does actually provide significant added value, and it is worthwhile keeping it around.
+
+![redis-edge-1](/redis-modules-arm-2.png)
+*Inside RedisEdge: moduels structure*
+
+Let's now take a look at each of the components of RedisEdge, to try to figure out what would it take to have them ported to ARM.
+First, **RedisTimeSeries**. 
+It's a simple C library, built with make. Not even a configure script. No problems there.
+Next is **RedisGears**. It's too a C library built with plain make, but it does use an embedded Python 3.7 interpreter, which is built from source. This requires running automake to generate a platform-specific makefile.
+Finally, **RedisAI**. It's a C library built with CMake, and it includes modular "engines" that allow abstraction and encapsulation of AI libraries like TensorFlow, PyTorch, and ONNXRuntime, in their C library form (most users typically use them in Python), with PyTorch and ONNXRuntime not officially supporting ARM.
+So the build requirements, as it seems, have deteriorated quickly. It went from building an innocent C library to compiling massive source bases with convoluted build systems.
+
+In the next sections, we will fit each components with its proper build method.
+
+## Building for ARM
+
+At this point, we'll pause and take stock of what is required to build software for ARM.
+The obvious way is to use an ARM-based device, like Raspberry Pi. Once set up, you'll be able to build and test in the most natural manner. 
+
+The testing medium is important: even though you can build almost every ARM software without a physical ARM device, there is no reliable way of testing the outcome in a reliable way without such a device, especially with non-standard devices. Therefore, as much as virtualized/emulated/containerized test may be useful, you should always test your software on its designated target device.

```

Comment:
Building on a Raspberry PI is cool, but wouldn't it make sense to do also describe doing ARM builds on the public cloud? e.g. using AWS a1 instances?

Reply:
I should have presented the next parts as well if they were ready by now. :-)
But yes, this is part of the next post, and eventually we're headed towards cross-building for ARM using Docker and Qemu user-mode emulation via binfmt_misc.
I didn't want this post be too long, too. So I settled for describing the physical ARM environment as it integrates neatly into the workstation environment (this is the major issue with AWS).
But again, once we'll have all parts ready, we'll have a much better view of the narrative.

Reply:
Sounds good. You may want to consider mentioning this in the post itself as a "teaser".